pkg/gadgets/traceloop/tracer/tracer.go

```diff
@@ -0,0 +1,483 @@
+//go:build linux
+// +build linux
+
+// Copyright 2019-2022 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracer
+
+// #include <linux/types.h>
+// #include "./bpf/traceloop.h"
+import "C"
+
+import (
+	"container/list"
+	"errors"
+	"fmt"
+	"math/rand"
+	"os"
+	"sync"
+	"unsafe"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
+	"github.com/cilium/ebpf/perf"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/traceloop/types"
+	eventtypes "github.com/inspektor-gadget/inspektor-gadget/pkg/types"
+	libseccomp "github.com/seccomp/libseccomp-golang"
+	log "github.com/sirupsen/logrus"
+)
+
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target ${TARGET} -cc clang traceloop ./bpf/traceloop.bpf.c -- -I./bpf/ -I../../../${TARGET}
+
+var (
+	useNullByteLength        uint64 = uint64(C.USE_NULL_BYTE_LENGTH)
+	useRetAsParamLength      uint64 = uint64(C.USE_RET_AS_PARAM_LENGTH)
+	useArgIndexAsParamLength uint64 = uint64(C.USE_ARG_INDEX_AS_PARAM_LENGTH)
+	paramProbeAtExitMask     uint64 = uint64(C.PARAM_PROBE_AT_EXIT_MASK)
+)
+
+var (
+	syscallEventTypeEnter uint8 = uint8(C.SYSCALL_EVENT_TYPE_ENTER)
+	syscallEventTypeExit  uint8 = uint8(C.SYSCALL_EVENT_TYPE_EXIT)
+	syscallEventTypeCont  uint8 = uint8(C.SYSCALL_EVENT_TYPE_CONT)
+)
+
+// This should match traceloop.h define SYSCALL_ARGS.
+var syscallArgs uint8 = 6
+
+var syscallsLock sync.Mutex
+var syscalls map[string]syscallDeclaration
```

Comment:
How about adding methods like this (I stuck with definition vs declaration in this example code):

```golang
type syscallDefinitions map[string]syscallDefinition

func (s *syscallDefinitions) getDefinition(syscallName string) (syscallDefinition, error) {
	syscall, ok := s[syscallName]
 	if !ok {
 		return syscallDefinition{}, fmt.Errorf("no syscall definition found for %q", syscallName)
 	}
 	return syscall, nil
}
```

and also:

```golang

func (s syscallDefinition) getParameterCount() uint8 {
	return uint8(len(s.params))
}

func (s syscallDefinition) getParameterName(paramNumber uint8) {
	if paramNumber >= len(s.params) {
		return "", fmt.Errorf("param number %d out of bounds for syscall %q", paramNumber, s.name)
	}
	return s.params[paramNumber].name, nil
}

```

This would get rid of the inline functions - which is a lot cleaner IMHO.